File: frontend/src/lib/api/reports.js
import { apiFetch } from './client.js';

/**
 * Company global ledger and analytics reporting API endpoints (FEAT-025, TODO-213/214).
 */

/**
 * @typedef {object} CompanyLedgerSummary
 * @property {string|null} period_from
 * @property {string|null} period_to
 * @property {number} new_projects_count
 * @property {number} new_clients_count
 * @property {number} new_services_count
 * @property {string} new_services_value
 * @property {string} total_invoiced
 * @property {string} total_collected
 * @property {string} total_due
 * @property {string} total_advance_balance
 */

/**
 * @typedef {object} CompanyLedgerTimelineItem
 * @property {string} period
 * @property {string} period_label
 * @property {number} new_projects
 * @property {number} new_clients
 * @property {number} new_services
 * @property {string} invoiced_amount
 * @property {string} collected_amount
 * @property {string} net_due_change
 */

/**
 * @typedef {object} CompanyLedgerProjectItem
 * @property {string} project_id
 * @property {string} short_id
 * @property {string} name
 * @property {string} client_id
 * @property {string} client_name
 * @property {string} status
 * @property {string} created_at
 * @property {number} services_count
 * @property {string} total_value
 * @property {string} total_invoiced
 * @property {string} total_paid
 * @property {string} balance_due
 */

/**
 * @typedef {object} CompanyLedgerClientItem
 * @property {string} client_id
 * @property {string} name
 * @property {string} client_type
 * @property {string} created_at
 * @property {number} active_projects_count
 * @property {number} total_services_count
 * @property {string} total_invoiced
 * @property {string} total_paid
 * @property {string} total_due
 * @property {string} advance_balance
 */

/**
 * @typedef {object} CompanyLedgerResponse
 * @property {CompanyLedgerSummary} summary
 * @property {CompanyLedgerTimelineItem[]} timeline
 * @property {CompanyLedgerProjectItem[]} projects
 * @property {CompanyLedgerClientItem[]} clients
 * @property {Record<string, any>} filters_applied
 */

/**
 * Fetch company-wide global ledger and operational analytics.
 *
 * @param {typeof fetch} fetchFn
 * @param {string} token
 * @param {object} [params]
 * @param {string} [params.date_from]
 * @param {string} [params.date_to]
 * @param {string} [params.client_id]
 * @param {string} [params.project_id]
 * @param {'month'|'day'} [params.granularity]
 * @param {'all'|'timeline'|'projects'|'clients'} [params.tab]
 * @returns {Promise<CompanyLedgerResponse>}
 */
export function getCompanyLedger(fetchFn, token, params = {}) {
	const sp = new URLSearchParams();
	if (params.date_from) sp.set('date_from', params.date_from);
	if (params.date_to) sp.set('date_to', params.date_to);
	if (params.client_id) sp.set('client_id', params.client_id);
	if (params.project_id) sp.set('project_id', params.project_id);
	if (params.granularity) sp.set('granularity', params.granularity);
	if (params.tab) sp.set('tab', params.tab);

	const qs = sp.toString();
	const path = qs ? `/tenant/reports/company-ledger?${qs}` : '/tenant/reports/company-ledger';
	return apiFetch(fetchFn, path, { token });
}

File: backend/app/schemas/reports.py
"""Schemas for company-wide reports and global analytics (FEAT-025, TODO-213)."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field


class CompanyLedgerSummary(BaseModel):
    """Executive KPI totals for the selected reporting window."""

    period_from: date | None = None
    period_to: date | None = None
    new_projects_count: int = 0
    new_clients_count: int = 0
    new_services_count: int = 0
    new_services_value: str = "0.00"
    total_invoiced: str = "0.00"
    total_collected: str = "0.00"
    total_due: str = "0.00"
    total_advance_balance: str = "0.00"


class CompanyLedgerTimelineItem(BaseModel):
    """Chronological time-bucketed rollup for operational and financial movement."""

    period: str  # e.g. "2026-09" or "2026-09-26"
    period_label: str  # e.g. "Sep 2026" or "26 Sep 2026"
    new_projects: int = 0
    new_clients: int = 0
    new_services: int = 0
    invoiced_amount: str = "0.00"
    collected_amount: str = "0.00"
    net_due_change: str = "0.00"


class CompanyLedgerProjectItem(BaseModel):
    """Project-level financial and operational summary."""

    project_id: uuid.UUID
    short_id: str
    name: str
    client_id: uuid.UUID
    client_name: str
    status: str
    created_at: datetime
    services_count: int = 0
    total_value: str = "0.00"
    total_invoiced: str = "0.00"
    total_paid: str = "0.00"
    balance_due: str = "0.00"


class CompanyLedgerClientItem(BaseModel):
    """Client-level cumulative billing and receivables summary."""

    client_id: uuid.UUID
    name: str
    client_type: str
    created_at: datetime
    active_projects_count: int = 0
    total_services_count: int = 0
    total_invoiced: str = "0.00"
    total_paid: str = "0.00"
    total_due: str = "0.00"
    advance_balance: str = "0.00"


class CompanyLedgerResponse(BaseModel):
    """Complete company global ledger & analytics response."""

    summary: CompanyLedgerSummary
    timeline: list[CompanyLedgerTimelineItem]
    projects: list[CompanyLedgerProjectItem]
    clients: list[CompanyLedgerClientItem]
    filters_applied: dict[str, Any] = Field(default_factory=dict)
